File: backend/internal/middleware/request_context.go
package middleware

import (
	"net/http"
	"runtime/debug"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"go.uber.org/zap"
)

const requestIDKey = "requestID"

// RequestContext assigns a request ID (from X-Request-ID header or a fresh UUID),
// echoes it back in the response header, and logs a structured access-log entry
// after the request completes. Log level is keyed on the response status code.
func RequestContext(logger *zap.Logger) gin.HandlerFunc {
	return func(c *gin.Context) {
		requestID := c.GetHeader("X-Request-ID")
		if requestID == "" {
			requestID = uuid.NewString()
		}

		c.Set(requestIDKey, requestID)
		c.Header("X-Request-ID", requestID)

		start := time.Now()
		c.Next()

		fields := []zap.Field{
			zap.String("request_id", requestID),
			zap.String("method", c.Request.Method),
			zap.String("path", c.FullPath()),
			zap.String("raw_path", c.Request.URL.Path),
			zap.Int("status_code", c.Writer.Status()),
			zap.Duration("duration", time.Since(start)),
			zap.String("client_ip", c.ClientIP()),
			zap.String("user_agent", c.Request.UserAgent()),
			zap.Int("response_bytes", c.Writer.Size()),
		}

		if userID, ok := c.Get("userID"); ok {
			fields = append(fields, zap.Any("user_id", userID))
		}

		if len(c.Errors) > 0 {
			fields = append(fields, zap.String("errors", c.Errors.String()))
		}

		switch {
		case c.Writer.Status() >= http.StatusInternalServerError:
			logger.Error("request completed", fields...)
		case c.Writer.Status() >= http.StatusBadRequest:
			logger.Warn("request completed", fields...)
		default:
			logger.Info("request completed", fields...)
		}
	}
}

// Recovery catches panics in downstream handlers, logs a full stack trace, and
// returns a 500 JSON error response so the server stays alive.
func Recovery(logger *zap.Logger) gin.HandlerFunc {
	return func(c *gin.Context) {
		defer func() {
			if recovered := recover(); recovered != nil {
				logger.Error("panic recovered",
					zap.String("request_id", GetRequestID(c)),
					zap.Any("panic", recovered),
					zap.ByteString("stack", debug.Stack()),
					zap.String("method", c.Request.Method),
					zap.String("path", c.Request.URL.Path),
				)
				c.AbortWithStatusJSON(http.StatusInternalServerError, gin.H{
					"error": gin.H{
						"code":    "SERVER_ERROR",
						"message": "Internal server error",
					},
				})
			}
		}()

		c.Next()
	}
}

// GetRequestID retrieves the request ID stored by RequestContext middleware.
// Returns an empty string if the middleware was not applied.
func GetRequestID(c *gin.Context) string {
	if v, ok := c.Get(requestIDKey); ok {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}

File: apps/api/src/middleware/request-context.ts
import type { Context, MiddlewareHandler } from "hono";
import type { Logger } from "pino";

export function requestContext(logger: Logger): MiddlewareHandler {
  return async (c, next) => {
    const requestId = c.req.header("X-Request-ID") ?? crypto.randomUUID();
    c.set("requestID", requestId);
    c.header("X-Request-ID", requestId);

    const start = Date.now();
    await next();

    const status = c.res.status;
    const durationMs = Date.now() - start;

    const fields: Record<string, unknown> = {
      request_id: requestId,
      method: c.req.method,
      path: c.req.routePath,
      raw_path: c.req.path,
      status_code: status,
      duration_ms: durationMs,
      user_agent: c.req.header("User-Agent") ?? "-",
    };

    const userId = c.get("userId");
    if (userId !== undefined) {
      fields.user_id = userId;
    }

    if (status >= 500) {
      logger.error(fields, "request completed");
    } else if (status >= 400) {
      logger.warn(fields, "request completed");
    } else {
      logger.info(fields, "request completed");
    }
  };
}

export function recovery(logger: Logger): MiddlewareHandler {
  return async (c, next) => {
    try {
      await next();
    } catch (err) {
      logger.error(
        {
          request_id: c.get("requestID") ?? "-",
          panic: String(err),
          stack: err instanceof Error ? err.stack : undefined,
          method: c.req.method,
          path: c.req.path,
        },
        "panic recovered",
      );
      return c.json(
        { error: { code: "SERVER_ERROR", message: "Internal server error" } },
        500,
      );
    }
  };
}

export function getRequestID(c: Context): string {
  return (c.get("requestID") as string) ?? "";
}
